### ISSUE TITLE
Sending block of codes behaves differently in Python script with cell and without cell

### ISSUE THREAD

**Issue by Fred-Wu** [posit-dev/positron#3907]
Positron: 2024.07.0-15
OS: MacOS

Without cell:

Only CTRL+ENTER works, and works when placing cursor anywhere in the block
<img width="339" alt="image" src="https://github.com/posit-dev/positron/assets/4111978/32fa2778-777d-40d8-8d0e-52147bf4cb36">

<img width="383" alt="image" src="https://github.com/posit-dev/positron/assets/4111978/6e220ba0-638b-4886-bb1a-43b2c37a5284">

However, with cell:
Both CTRL+ENTER and SHIFT+ENTER work, but behave differently

SHIFT+ENTER works when placing the cursor anywhere in the block, but NOT CTRL+ENTER (it only works when placing the cursor in the beginning of the block)

CTRL+ENTER

<img width="385" alt="image" src="https://github.com/posit-dev/positron/assets/4111978/8598be51-a839-4f8b-be16-cd93bd6df664">

<img width="385" alt="image" src="https://github.com/posit-dev/positron/assets/4111978/d9d0d4df-ab26-4309-8577-4e93c96a5c5e">

**Comment by juliasilge** [comment]
Yep, similar to #3908, this is currently behaving as designed/expected but we are happy to hear feedback about what you were expecting or would prefer.

<kbd>Shift</kbd>+<kbd>Enter</kbd> is a keyboard shortcut specific to code cells to run the whole code cell. There is a whole set of keyboard shortcuts specific to these kinds of cells:

<img width="850" alt="Screenshot 2024-07-07 at 8 36 29 PM" src="https://github.com/posit-dev/positron/assets/12505835/321d0a31-f35b-4bac-87cc-116acf6660d3">

Similar to what I outline in #3908, the keyboard shortcut to execute the current statement looks for the immediate current complete statement. We do have some open questions about Python statement detection in #3464 if you'd like to add your experiences and preferences there.

**Comment by petetronic** [comment]
It looks like though that there is an issue with code block detection? Opening for RC?

**Comment by seeM** [comment]
@petetronic yes it looks like there's a bug with our fallback regex-based detector. Here's a simple repro:

1. Write this to a file:

    ```python
    x = # this line is to intentionally create a syntax error so that we fall back to the regex-based detector
    y = [
        0,
    ]
    ```
2. Place the cursor on the `0,` line and press Cmd+Enter.
3. Only the line `0,` gets sent to the console.

It should send the entire `y` definition to the console.

Note that it works as expected without the `x =` syntax error.

#### Possible solutions

We can try to look into fixing this case for the regex-based detector. I wonder though if we should consider moving the AST-based detector to a parser that supports syntax errors (originally raised this [here](https://github.com/posit-dev/positron/issues/1260#issuecomment-1812680020)).

I haven't looked into all of the options, but I know that [`parso`](https://github.com/davidhalter/parso) (used by Jedi and maintained by the same person) does support error recovery.

**Comment by austin3dickey** [comment]
quick update on this:

1. Running Ctrl/Cmd + Enter inside a code cell does run the whole cell today. It doesn't just send one line like in the original report.
2. Wasim's reprex about what happens when there's a syntax error still fails today, so I'll re-title this issue to primarily deal with that.
3. I thought this could get fixed by #10305 but it's completely unrelated since statement range isn't part of the LSP. I'll move this to next quarter instead.

**Comment by juliasilge** [comment]
Related to #12095 in terms of syntax errors and the statement range provider
